cpp/src/arrow/flight/sql/driver.cc

```diff
@@ -0,0 +1,1877 @@
+  if (!statement->private_data) return ADBC_STATUS_INVALID_STATE;
+  auto* ptr =
+      reinterpret_cast<std::shared_ptr<FlightSqlStatementImpl>*>(statement->private_data);
+  return (*ptr)->GetParameterSchema(schema, error);
+}
+
+AdbcStatusCode FlightSqlStatementNew(struct AdbcConnection* connection,
+                                     struct AdbcStatement* statement,
+                                     struct AdbcError* error) {
+  auto* ptr = reinterpret_cast<std::shared_ptr<FlightSqlConnectionImpl>*>(
+      connection->private_data);
+  auto impl = std::make_shared<FlightSqlStatementImpl>(*ptr);
+  statement->private_data = new std::shared_ptr<FlightSqlStatementImpl>(impl);
+  return ADBC_STATUS_OK;
+}
+
+AdbcStatusCode FlightSqlStatementPrepare(struct AdbcStatement* statement,
+                                         struct AdbcError* error) {
+  if (!statement->private_data) return ADBC_STATUS_INVALID_STATE;
+  auto* ptr =
+      reinterpret_cast<std::shared_ptr<FlightSqlStatementImpl>*>(statement->private_data);
+  return (*ptr)->Prepare(error);
+}
+
+AdbcStatusCode FlightSqlStatementRelease(struct AdbcStatement* statement,
+                                         struct AdbcError* error) {
+  if (!statement->private_data) return ADBC_STATUS_INVALID_STATE;
+  auto* ptr =
+      reinterpret_cast<std::shared_ptr<FlightSqlStatementImpl>*>(statement->private_data);
+  auto status = (*ptr)->Close(error);
+  delete ptr;
+  statement->private_data = nullptr;
+  return status;
+}
+
+AdbcStatusCode FlightSqlStatementSetOption(struct AdbcStatement* statement,
+                                           const char* key, const char* value,
+                                           struct AdbcError* error) {
+  if (!statement->private_data) return ADBC_STATUS_INVALID_STATE;
+  auto* ptr =
+      reinterpret_cast<std::shared_ptr<FlightSqlStatementImpl>*>(statement->private_data);
+  return (*ptr)->SetOption(key, value, error);
+}
+
+AdbcStatusCode FlightSqlStatementSetSqlQuery(struct AdbcStatement* statement,
+                                             const char* query, struct AdbcError* error) {
+  if (!statement->private_data) return ADBC_STATUS_INVALID_STATE;
+  auto* ptr =
+      reinterpret_cast<std::shared_ptr<FlightSqlStatementImpl>*>(statement->private_data);
+  return (*ptr)->SetSqlQuery(query, error);
+}
+
+AdbcStatusCode FlightSqlStatementSetSubstraitPlan(struct AdbcStatement* statement,
+                                                  const uint8_t* plan, size_t length,
+                                                  struct AdbcError* error) {
+  if (!statement->private_data) return ADBC_STATUS_INVALID_STATE;
+  auto* ptr =
+      reinterpret_cast<std::shared_ptr<FlightSqlStatementImpl>*>(statement->private_data);
+  return (*ptr)->SetSubstraitPlan(plan, length, error);
+}
+}  // namespace
+}  // namespace arrow::flight::sql
+
+AdbcStatusCode AdbcDatabaseInit(struct AdbcDatabase* database, struct AdbcError* error) {
```

Comment:
They're at the top level since one of the original goals was to be able to link to the driver and use it directly with the same API as the driver manager.

That's why the declarations are repeated twice. If you link directly, you get these definitions. If you link to the driver manager, it provides the definitions of these symbols. It then loads this library with RTLD_LOCAL (so the driver's symbols don't get made available) and uses dlopen to get everything. (It would be best to also RTLD_DEEPBIND, but I found this breaks with AddressSanitizer.)

Reply:
> They're at the top level since one of the original goals was to be able to link to the driver and use it directly with the same API as the driver manager.

I'm afraid I don't understand this use case. Especially as we probably want to provide a higher-level C++ API around ADBC at some point? Whoever links directly against `libflight_sql` is probably ok with writing C++ code.

In any case, it seems that multiple drivers loaded in the same process would have clashing top-level symbols.

Reply:
This applies to other drivers, too. If you only want to use say SQLite, you can then link to libadbc_driver_sqlite directly, but use the ADBC API. Admittedly it is a unique use case. (Honestly, I thought you originally suggested keeping the driver/driver manager API the same :slightly_smiling_face:)

At least as implemented, RTLD_LOCAL means that dynamic linker won't let other things find the clashing symbol. So it would be OK to load multiple drivers.

Reply:
(But, this is why the driver has a namespaced version of each symbol for use in the dynamic entry point, in case that is not true.)